webforces/server/tests.py

```diff
@@ -44,34 +43,219 @@ def test_cant_add_iden_users(self):
+        user2 = User(0, "LOGIN_USER2", "FN_USER2", "SN_USER2", "MN_USER2", [])
+        user1 = (self.core.db.addUser(user1))[1]
+        user2 = (self.core.db.addUser(user2))[1]
+        alg1 = Algorithm(0, "TITLE_ALG1", user1.user_id, "SOURCE_ALG1", [], 0)
+        alg2 = Algorithm(0, "TITLE_ALG2", user1.user_id, "SOURCE_ALG2", [], 0)
+        alg3 = Algorithm(0, "TITLE_ALG3", user2.user_id, "SOURCE_ALG3", [], 0)
+        status1, alg1 = self.core.db.addAlg(alg1)
+        status2, alg2 = self.core.db.addAlg(alg2)
+        status3, alg3 = self.core.db.addAlg(alg3)
         self.assertEqual(status1, DBStatus.s_ok)
         self.assertEqual(status2, DBStatus.s_ok)
-        self.assertEqual(user1, user2)
+        self.assertEqual(status3, DBStatus.s_ok)
+        self.assertEqual(alg1.alg_id, 1)
+        self.assertEqual(alg2.alg_id, 2)
+        self.assertEqual(alg3.alg_id, 1)
+        self.assertEqual(self.core.db._getNextID("1u"), 3)
+        self.assertEqual(self.core.db._getNextID("2u"), 2)
+        self.assertEqual(self.core.db._getNextID("1u1a"), 1)
+        self.assertEqual(self.core.db._getNextID("1u2a"), 1)
+        self.assertEqual(self.core.db._getNextID("2u1a"), 1)
+
+    def test_cant_add_alg_with_incorrect_user_id(self):
+        user = User(0, "LOGIN_USER", "FN_USER", "SN_USER", "MN_USER", [])
+        user = (self.core.db.addUser(user))[1]
+        alg = Algorithm(0, "TITLE_ALG", user.user_id + len("typo"), "SOURCE_ALG", [], 0)
+        status, alg = self.core.db.addAlg(alg)
+        self.assertEqual(status, DBStatus.s_data_issue)
+        self.assertEqual(alg.alg_id, -100)
+        self.assertEqual(self.core.db._getNextID("1u"), 1)
+        self.assertEqual(self.core.db._getNextID("1u1a"), DBStatus.s_data_issue)
 
-    def test_cant_get_user_by_incorrect_id(self):
+    def test_cant_add_iden_algs(self):
         user1 = User(0, "LOGIN_USER1", "FN_USER1", "SN_USER1", "MN_USER1", [])
-        status1, user1 = self.core.db.addUser(user1)
-        status2, user2 = self.core.db.getUserByID(user1.user_id + 1)
+        user2 = User(0, "LOGIN_USER2", "FN_USER2", "SN_USER2", "MN_USER2", [])
+        user1 = (self.core.db.addUser(user1))[1]
+        user2 = (self.core.db.addUser(user2))[1]
+        alg1 = Algorithm(0, "TITLE_ALG1", user1.user_id, "SOURCE_ALG1", [], 0)
+        alg2 = Algorithm(0, "TITLE_ALG1", user1.user_id, "SOURCE_ALG2", [], 0)
+        alg3 = Algorithm(0, "TITLE_ALG1", user2.user_id, "SOURCE_ALG3", [], 0)
+        status1, alg1 = self.core.db.addAlg(alg1)
+        status2, alg2 = self.core.db.addAlg(alg2)
+        status3, alg3 = self.core.db.addAlg(alg3)
         self.assertEqual(status1, DBStatus.s_ok)
         self.assertEqual(status2, DBStatus.s_data_issue)
-        self.assertEqual(user2.user_id, -100)
+        self.assertEqual(status3, DBStatus.s_data_issue)
+        self.assertEqual(alg1.alg_id, 1)
+        self.assertEqual(alg2.alg_id, -100)
```

Comment:
Btw, noticed that this constant looks pretty weird. What do you think about adding some constant (or enumeration if there're several constants) for invalid ID(s).

Reply:
Done.